pkg/virt-controller/watch/vm.go

```diff
@@ -2025,6 +1921,130 @@ func (c *VMController) processFailureCondition(vm *virtv1.VirtualMachine, vmi *v
+		stateChange := vm.Status.StateChangeRequests[0]
+		switch stateChange.Action {
+		case virtv1.StopRequest:
+			if vmi == nil {
+				// because either the VM or VMI informers can trigger processing here
+				// double check the state of the cluster before taking action
+				_, err := c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Get(vm.GetName(), &v1.GetOptions{})
+				if err != nil && errors.IsNotFound(err) {
+					// If there's no VMI, then the VMI was stopped, and the stopRequest can be cleared
+					log.Log.Object(vm).V(4).Infof("No VMI. Clearing stop request")
+					clearChangeRequest = true
+				}
+			} else {
+				if stateChange.UID == nil {
+					// It never makes sense to have a request to stop a VMI that doesn't
+					// have a UUID associated with it. This shouldn't be possible -- but if
+					// it occurs, clear the stopRequest because it can't be acted upon
+					log.Log.Object(vm).Errorf("Stop Request has no UID.")
+					clearChangeRequest = true
+				} else if *stateChange.UID != vmi.UID {
+					// If there is a VMI, but the UID doesn't match, then it
+					// must have been previously stopped, so the stopRequest can be cleared
+					log.Log.Object(vm).V(4).Infof("VMI's UID doesn't match. clearing stop request")
+					clearChangeRequest = true
+				}
+			}
+		case virtv1.StartRequest:
+			// If the current VMI is running, then it has been started.
+			if vmi != nil {
+				log.Log.Object(vm).V(4).Infof("VMI exists. clearing start request")
+				clearChangeRequest = true
+			}
+		}
+	}
+
+	return clearChangeRequest
+}
+
+func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
+	if len(vm.Status.VolumeRequests) > 0 {
+		volumeMap := make(map[string]virtv1.Volume)
+		diskMap := make(map[string]virtv1.Disk)
+
+		for _, volume := range vm.Spec.Template.Spec.Volumes {
+			volumeMap[volume.Name] = volume
+		}
+		for _, disk := range vm.Spec.Template.Spec.Domain.Devices.Disks {
+			diskMap[disk.Name] = disk
+		}
+
+		tmpVolRequests := vm.Status.VolumeRequests[:0]
+		for _, request := range vm.Status.VolumeRequests {
+
+			var added bool
+			var volName string
+
+			removeRequest := false
+
+			if request.AddVolumeOptions != nil {
+				volName = request.AddVolumeOptions.Name
+				added = true
+			} else if request.RemoveVolumeOptions != nil {
+				volName = request.RemoveVolumeOptions.Name
+				added = false
+			}
+
+			_, volExists := volumeMap[volName]
+			_, diskExists := diskMap[volName]
+
+			if added && volExists && diskExists {
+				removeRequest = true
+			} else if !added && !volExists && !diskExists {
+				removeRequest = true
+			}
+
+			if !removeRequest {
+				tmpVolRequests = append(tmpVolRequests, request)
+			}
+		}
+		vm.Status.VolumeRequests = tmpVolRequests
+	}
+}
+
+func (c *VMController) syncChangeRequest(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance, key string, dataVolumes []*cdiv1.DataVolume) (syncError, error) {
```

Comment:
Sorry for countering @marceloamaral proposal, but I'm not sure `syncChangeRequest` is the best name here too. This terminology incorrectly suggests that this operates on the VM's StateChangeRequests, which is not the case.

TBH, it's not easy to come up with a name for this function, since it doesn't really encapsulate a well-defined functionality. I think we'll need (at some point) to return to `execute()` and see how we can refactor it into more cohesive pieces of code, including this. In the meantime, I'd say that extracting this function as-it-is doesn't make the flow more readable, but rather only more scattered, so I'd just move it it back into `execute()` 🤷🏼‍♂️

Reply:
I agree that `syncChangeRequest` is not the best name, it can be misleading as it is actually executing the changeRequest....

Maybe take all the part related to the `syncErr` variable, move to the function and call the function just `sync` or `executeSync`.

We probably need help from @rmohr or @davidvossel

Reply:
i'd just call it `sync()`. That's consistent with the vmi and other controllers